ext/gd/gd.c

```diff
@@ -4273,114 +4373,44 @@ static gdIOCtx *create_stream_context(php_stream *stream, int close_stream) {
 	return ctx;
 }
 
-static gdIOCtx *create_output_context(void) {
-	gdIOCtx *ctx = ecalloc(1, sizeof(gdIOCtx));
-
-	ctx->putC = _php_image_output_putc;
-	ctx->putBuf = _php_image_output_putbuf;
-	ctx->gd_free = _php_image_output_ctxfree;
+static gdIOCtx *create_output_context(zval *to_zval, uint32_t arg_num) {
+	gdIOCtx *ctx;
 
-	return ctx;
-}
+	if (to_zval != NULL) {
+		php_stream *stream;
+		int close_stream = 1;
 
-static void _php_image_output_ctx(INTERNAL_FUNCTION_PARAMETERS, int image_type, char *tn)
-{
-	zval *imgind;
-	zend_long quality = -1, basefilter = -1, speed = -1;
-	gdImagePtr im;
-	gdIOCtx *ctx = NULL;
-	zval *to_zval = NULL;
+		if (Z_TYPE_P(to_zval) == IS_RESOURCE) {
+			php_stream_from_zval_no_verify(stream, to_zval);
+			if (stream == NULL) {
+				return NULL;
+			}
+			close_stream = 0;
+		} else if (Z_TYPE_P(to_zval) == IS_STRING) {
+			if (CHECK_ZVAL_NULL_PATH(to_zval)) {
+				zend_argument_type_error(arg_num, "must not contain null bytes");
+				return NULL;
+			}
```

Comment:
Just for sanity check, can you add ``ZEND_ASSERT(arg_num != 0)``, possibly just after the `if` where we check that `to_zval` is not `NULL`